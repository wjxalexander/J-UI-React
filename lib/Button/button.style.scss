@import "../_helper.scss";

%default-style {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
  color: $default_white;
  font-family: inherit;
  text-decoration: none;
  border-style: none;
  min-width: 80px;
  max-width: 280px;
  border-radius: 2px;
  //min-height: 72px;
  //height: auto;
  height: 32px;
  font-size: $default_fontsize;
  font-weight: $default_text_title_fontweight;
  cursor: pointer;
  outline: none;
}

.j-ui-button {
  &-default-wrapper {
    @extend %default-style;
    background-color: $default_color;
    border: #000;
    &:hover,
    &:active {
      background-color: $button_hover_color;
    }
  }
;

  &-custom-wrapper {
    @extend %default-style;

    border: 1px solid $custom-border-color;
    background-color: #fff;
    color: #000;

    &:hover,
    &:active {
      background-color: rgb(243, 242, 241);
      color: rgb(32, 31, 30);
    }
  }
;

  &-secondaryText {
    flex-grow: 1;
    font-weight: 400;
    font-size: 12px;
  }
}

.j-ui-button-icon {
  font-size: $default_fontsize;
  margin: 0 4px;
}

button:disabled {
  color: rgba(0, 0, 0, 0.25);
  background-color: #f5f5f5;
  border-color: #d9d9d9;
  cursor: not-allowed;
  &:hover{
    color: rgba(0, 0, 0, 0.25);
    background-color: #f5f5f5;
  }
}