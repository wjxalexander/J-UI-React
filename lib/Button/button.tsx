import * as React from 'react';
import {rootClass} from "../utils/classFactory";
import Icon, {IconProps} from "../Icon/icon";
import "./button.style.scss"
import cls from "classnames"

export interface buttonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  className?: string;
  title?: string;
  buttonType: "default" | "custom";
  iconProps?: IconProps;
  secondaryText?: string;
}

const classGenerator = rootClass('j-ui-button');

const Button: React.FunctionComponent<buttonProps> = (props: buttonProps) => {
  const {buttonType,iconProps,secondaryText,...restProps} = props;
  return (
    <button {...restProps} className={classGenerator(`${buttonType}-wrapper`)}>
      {iconProps ?
        <Icon {...iconProps} className={cls('j-ui-button-icon', iconProps.className)}/> : null}
      <div className={classGenerator(`${buttonType}-title`)}>
        {props.title || "button"}
      </div>
      {secondaryText ?
        <div className={classGenerator(`${buttonType}-secondaryText`)}>{secondaryText}</div> : null}
    </button>)
};
 Button.defaultProps = {
   buttonType: 'default'
 }
export default Button;