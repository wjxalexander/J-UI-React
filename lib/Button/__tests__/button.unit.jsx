import React from 'react';
import { mount} from 'enzyme';
import Button from '../button';
import renderer from 'react-test-renderer'

describe('Button', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Button title="test" buttonType="default"/>)
      .toJSON(); //虚拟dom是个对象
    expect(tree).toMatchSnapshot()
  });
  it('renders correctly withIcon', () => {
    const tree = renderer
      .create(<Button title="test" buttonType="default" iconProps={{iconName: "accept", USEMsFabricIcon: true}}/>)
        .toJSON(); //虚拟dom是个对象
        expect(tree).toMatchSnapshot()
      });
  it('renders correctly withoutTitle', () => {
    const tree = renderer
      .create(<Button buttonType="default" iconProps={{iconName: "accept", USEMsFabricIcon: true}}/>)
      .toJSON(); //虚拟dom是个对象
    expect(tree).toMatchSnapshot()
  });
  it('renders correctly with secondaryTitle', () => {
    const tree = renderer
      .create(<Button buttonType="default" iconProps={{iconName: "accept", USEMsFabricIcon: true}} secondaryText="test1111"/>)
      .toJSON(); //虚拟dom是个对象
    expect(tree).toMatchSnapshot()
  });
  it("onClick Button", () => {
    const fakeFn = jest.fn();
    const component = mount(<Button title="test" buttonType="default" onClick={fakeFn}/>);
    component.find('button').simulate('click')
    expect(fakeFn).toBeCalled()
  });
  it("onClick Button disabled", () => {
    const fakeFn = jest.fn();
    const component = mount(<Button title="test" buttonType="default" disabled/>);
    component.find('button').simulate('click')
    expect(fakeFn).not.toBeCalled()
  });
  });
