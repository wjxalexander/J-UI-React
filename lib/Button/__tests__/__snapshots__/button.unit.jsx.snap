// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<button
  className="j-ui-button-default-wrapper"
  title="test"
>
  <div
    className="j-ui-button-default-title"
  >
    test
  </div>
</button>
`;

exports[`Button renders correctly with secondaryTitle 1`] = `
<button
  className="j-ui-button-default-wrapper"
>
  <i
    className="ms-Icon ms-Icon--accept J-UI-Icon j-ui-button-icon"
  />
  <div
    className="j-ui-button-default-title"
  >
    button
  </div>
  <div
    className="j-ui-button-default-secondaryText"
  >
    test1111
  </div>
</button>
`;

exports[`Button renders correctly withIcon 1`] = `
<button
  className="j-ui-button-default-wrapper"
  title="test"
>
  <i
    className="ms-Icon ms-Icon--accept J-UI-Icon j-ui-button-icon"
  />
  <div
    className="j-ui-button-default-title"
  >
    test
  </div>
</button>
`;

exports[`Button renders correctly withoutTitle 1`] = `
<button
  className="j-ui-button-default-wrapper"
>
  <i
    className="ms-Icon ms-Icon--accept J-UI-Icon j-ui-button-icon"
  />
  <div
    className="j-ui-button-default-title"
  >
    button
  </div>
</button>
`;
